### ISSUE TITLE
psutil version used by Telemetry needs to be updated

### ISSUE THREAD

**Issue by zeptonaut** [catapult-project/catapult#2739]
It seems like the CPU tracing agent that's supposed to capture the output of `top` periodically throughout a test is failing to do so due to permission errors.

[Here's a link (Google internal) to such a failure.](https://build.chromium.org/p/tryserver.chromium.perf/builders/winx64_high_dpi_perf_bisect/builds/11/steps/Performance%20Test%20%28With%20Patch%29%201%20of%201/logs/stdio)

It looks like:

```
@@@STEP_LOG_LINE@Captured Output@Traceback (most recent call last):@@@
@@@STEP_LOG_LINE@Captured Output@  File "C:\b\c\b\winx64_high_dpi_perf_bisect\src\third_party\catapult\telemetry\telemetry\internal\platform\tracing_agent\cpu_tracing_agent.py", line 108, in GetProcesses@@@
@@@STEP_LOG_LINE@Captured Output@    'name': p.name,@@@
@@@STEP_LOG_LINE@Captured Output@  File "C:\b\depot_tools\python276_bin\lib\site-packages\psutil\_common.py", line 48, in __get__@@@
@@@STEP_LOG_LINE@Captured Output@    ret = self.func(instance)@@@
@@@STEP_LOG_LINE@Captured Output@  File "C:\b\depot_tools\python276_bin\lib\site-packages\psutil\__init__.py", line 341, in name@@@
@@@STEP_LOG_LINE@Captured Output@    name = self._platform_impl.get_process_name()@@@
@@@STEP_LOG_LINE@Captured Output@  File "C:\b\depot_tools\python276_bin\lib\site-packages\psutil\_psmswindows.py", line 190, in wrapper@@@
@@@STEP_LOG_LINE@Captured Output@    return fun(self, *args, **kwargs)@@@
@@@STEP_LOG_LINE@Captured Output@  File "C:\b\depot_tools\python276_bin\lib\site-packages\psutil\_psmswindows.py", line 222, in get_process_name@@@
@@@STEP_LOG_LINE@Captured Output@    return os.path.basename(self.get_process_exe())@@@
@@@STEP_LOG_LINE@Captured Output@  File "C:\b\depot_tools\python276_bin\lib\site-packages\psutil\_psmswindows.py", line 194, in wrapper@@@
@@@STEP_LOG_LINE@Captured Output@    raise AccessDenied(self.pid, self._process_name)@@@
@@@STEP_LOG_LINE@Captured Output@AccessDenied: (pid=652)@@@
```

The job associated with that log was trying to run the battor.power_cases benchmark.

/cc @wangzhen127 @Alex319 @randalnephew

**Comment by anniesullie** [comment]
Do you have any more details on how often this fails? Has it always failed, did it start failing recently, is it just flaky?

**Comment by zeptonaut** [comment]
Seems to be working reliably now. I'll reopen this if I run into problems again.

**Comment by zeptonaut** [comment]
Reopening this. The underlying problem has not been fixed, even if the tracing agent itself has been disabled on all bots. We would like to reenable it to help reduce noise in our BattOr power tests.

**Comment by zeptonaut** [comment]
So the main problem here is that old versions of psutil on Windows require significantly more permissions than we give Telemetry. New versions of psutil require fewer permissions. Unfortunately, we currently inherit our version of psutil from depot_tools, which is very out of date (they use a version from 2013). When I asked them to roll the version, they said that they would really prefer if we just used our own version for Telemetry in order to prevent versioning hell down the road.

The answer here seems to be to include psutil in third_party and to use that version of the library instead of the default one that ships with depot_tools. This is slightly complicated by the fact that psutil also relies on some compiled C/C++ libraries (included in distributions as `.so` files). I believe that if we check in `.so` files from each platform (similar to what we do with vinn), we should be okay.

In the long run, I think that we may want to have a `requirements.txt` file for Telemetry that specifies exactly what dependencies we need. It may also not be a bad idea to use `virtualenv` to prevent us from pulling in more global libraries like we accidentally did here.

**Comment by anniesullie** [comment]
@Apeliotes @nedn @ehanley324 sounds like a job for dependency manager?

**Comment by zeptonaut** [comment]
In this case, what's the advantage of using the dependency manager rather than just keeping the files directly in the repo? Each of the files is suffixed with its platform (`_psutil_[osx,win,posix].so`).

**Comment by zeptonaut** [comment]
I'm honestly not sure how portable shared libraries (`.so` files) are. I'm going to spend a few minutes looking into that before deciding whether this is a reasonable approach.

**Comment by nedn** [comment]
dependency_manager is our way of getting around the problem of checking big binaries file into git tree is bad, because they don't diff well & blows up the checkout side.

But for the case of ps_util, I don't think it's a good idea to use dependency_manager yet, because psutil is a python module, whereas dependency_manager is a runtime thing that fetch binaries. We resolve the module import during import time, which is earlier than when dependecy_manager run.

I think you should push harder for infra to upgrade their version of psutil. The fact that telemetry is using a different lib from infra has bite us with crbug.com/656224, so I prefer if we don't use a copy of library which is not the same as the one infra provides.

**Comment by anniesullie** [comment]
Where is the link to the bug requesting an update to depot_tools psutil?

**Comment by nedn** [comment]
crbug.com/655806

**Comment by zeptonaut** [comment]
The reason I'm hesitant to push them is mostly because I think their point is valid, though. We really only care about them upgrading because it's hard to provide the package for ourselves, which is a problem. Their argument of "if you want a specific version of a library, use `virtualenv`+`pip`" is more consistent with best practices than my argument of "upgrade the global version" was. I just don't feel like I have a lot of ground to stand on for that argument.

**Comment by zeptonaut** [comment]
Closing this as WontFix: this ended up being *much* more complicated than I expected. Basically, we're using the version of psutil that's supplied by depot_tools, which is very much discouraged. They suggest using a package manager and virtualenv with our own version of psutil if we want to use an updated version, which is a much larger change than just updating depot_tools psutil. Overall, we just ended up implementing the CPU tracing agent another way.
